### ISSUE TITLE
Can Guice do DI without the @Inject annotation on constructors without real side-effects?

### ISSUE THREAD

**Issue by rajanadar** [google/guice#1079]
This is a question (rather than a bug or anything) for the Guice team. and a long one at that, so please grab a coffee. :)

this is based on working on the full Spring stack for years, Spring boot and Guice recently, and Unity/Autofac on the .NET side for years.. (so a fair bit of DI engines)

- Most systems have a majority of scenarios (we have more than 95%) where 1 interface has just 1 concrete class. For these,
    - In Spring, we annotate the concrete with @Component and @Inject (or @Autowired) in constructors.
    - In Guice, we annotate the concrete with just @Inject and leave the bindings to the Module. (outside of the concrete class)
- With Constructor injection as the prevalent pattern (again more than 95% for us), **is there really a TRUE need for the @Inject annotation on top of a constructor for Guice? Can Guice work without any side-effects if we don't annotate concrete class constructors with @Inject?**

The aim is that, for folks using purely constructor based DI, their classes are 100% clean of any DI annotations. All DI happens outside of it. (Abstract Module)

This is something that tools like Autofac (for .NET) support. They don't need the classes to be tampered with.

More details:

- When we register the bindings in the guice modules, we tell Guice that, "hey, when someone requests IBillingProcessor, give them "BillingProcessor" class." 
- Since it is "turtles all the way down", Guice tries to keep resolving it, till the instance can be constructed.     
    - As part of this, WHEN IT COMES TO CONSTRUCTORs, Guice could totally keep constructing the concrete classes, irrespective of whether the @Inject annotation is present.
    - Are there scenarios where folks have multiple constructors, and don't want some of them to participate in DI? I would think, teams will be happy to let the DI select the best matched constructor. (if there are many)

For property based DI and multiple concrete classes, such scenarios can warrant some sort of annotation since Guice (or any framework) has no way of knowing otherwise. So an annotation is totally fine there.

But for the normal constructor based DI scenarios, can Abstract Module encapsulate all of DI logic, and keep the real classes/interfaces clean of any DI annotations/knowledge?

Thanks for reading through.

**Comment by mcculls** [comment]
By default Guice does not require you to annotate your public constructor
with @Inject - if there's only one such public constructor then it will
automatically pick that one. Even when there are multiple constructors then
you can use a constructor binding to tell Guice which one to use and avoid
the ambiguity:

https://google.github.io/guice/api-docs/latest/javadoc/com/google/inject/binder/LinkedBindingBuilder.html#toConstructor-java.lang.reflect.Constructor-

Using @Inject is still recommended for code under your control because it
documents that class was written with injection in mind.

You can also tell Guice to enforce the existence of @Inject on constructors
if you wish:

https://google.github.io/guice/api-docs/latest/javadoc/com/google/inject/Binder.html#requireAtInjectOnConstructors--

--
Cheers, Stuart


On 4 Feb 2017 06:36, "Raja Nadar" <notifications@github.com> wrote:

This is a question (rather than a bug or anything) for the Guice team. and
a long one at that, so please grab a coffee. :)

this is based on working on the full Spring stack for years, Spring boot
and Guice recently, and Unity/Autofac on the .NET side for years.. (so a
fair bit of DI engines)

   - Most systems have a majority of scenarios (we have more than 95%)
   where 1 interface has just 1 concrete class. For these,
      - In Spring, we annotate the concrete with @Component
      <https://github.com/Component> and @Inject <https://github.com/Inject>
      (or @Autowired <https://github.com/Autowired>) in constructors.
      - In Guice, we annotate the concrete with just @Inject
      <https://github.com/Inject> and leave the bindings to the Module.
      (outside of the concrete class)
   - With Constructor injection as the prevalent pattern (again more than
   95% for us), *is there really a TRUE need for the @Inject
   <https://github.com/Inject> annotation on top of a constructor for Guice?
   Can Guice work without any side-effects if we don't annotate concrete class
   constructors with @Inject <https://github.com/Inject>?*

The aim is that, for folks using purely constructor based DI, their classes
are 100% clean of any DI annotations. All DI happens outside of it.
(Abstract Module)

This is something that tools like Autofac (for .NET) support. They don't
need the classes to be tampered with.

More details:

   - When we register the bindings in the guice modules, we tell Guice
   that, "hey, when someone requests IBillingProcessor, give them
   "BillingProcessor" class."
   - Since it is "turtles all the way down", Guice tries to keep resolving
   it, till the instance can be constructed.
      - As part of this, WHEN IT COMES TO CONSTRUCTORs, Guice could totally
      keep constructing the concrete classes, irrespective of whether the
      @Inject <https://github.com/Inject> annotation is present.
      - Are there scenarios where folks have multiple constructors, and
      don't want some of them to participate in DI? I would think,
teams will be
      happy to let the DI select the best matched constructor. (if
there are many)

For property based DI and multiple concrete classes, such scenarios can
warrant some sort of annotation since Guice (or any framework) has no way
of knowing otherwise. So an annotation is totally fine there.

But for the normal constructor based DI scenarios, can Abstract Module
encapsulate all of DI logic, and keep the real classes/interfaces clean of
any DI annotations/knowledge?

Thanks for reading through.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/google/guice/issues/1079>, or mute the thread

**Comment by sblommers** [comment]
Or you could just use @Provides in you (abstract)module for your specific concrete class implementing that interface.

**Comment by yuranos** [comment]
@mcculls , it didn't work that way for me. I've got [Missing Constructor](https://github.com/google/guice/wiki/MISSING_CONSTRUCTOR) exception when I removed @Inject. My classes had exactly one public constructor and yet it's not enough if that constructor has arguments.
In any case, @Provides work like a charm if the intention is to keep classes Guice-unaware and configure everything programmatically in the modules.
